### PR TITLE
#165415819 Setup postgresql database

### PR DESCRIPTION
**Pull request by Nerocodes**
### Description of Task to be completed?
Setup postgresql database

### Any background context you want to provide?
NA

### What are the relevant pivotal tracker stories?
[#165415819](https://www.pivotaltracker.com/story/show/165415819)

### Screenshots (if appropriate)
NA

### Questions:
NA

### CODE REVIEW

File: api/package.json [hunk 1]
```diff
@@ -7,10 +7,19 @@
...
-    "start": "node ./build/index.js",
+    "start": "set NODE_ENV='PROD' && node ./build/index.js",
     "dev-start": "nodemon --exec babel-node ./src/index.js",
-    "test": "nyc mocha --require @babel/register ./src/test/ --exit || true",
-    "coverage": "nyc npm test && nyc report --reporter=text-lcov | coveralls"
+    "test": "set NODE_ENV='TEST' && npm run migrations && npm run seeders && nyc mocha --timeout=20000 --require @babel/register ./src/test/ --exit || true",
+    "singleTest": "set NODE_ENV='TEST' && nyc mocha --timeout=20000 --require @babel/register ./src/test/account.test.js --exit || true",
+    "coverage": "nyc npm test && nyc report --reporter=text-lcov | coveralls",
+    "migrateUser": "babel-node ./src/database/migrations/user createUserTable",
+    "migrateAccount": "babel-node ./src/database/migrations/account createAccountTable",
```
Comment (ucheg6) [inline api/package.json]: why are you migrating the tables differently?

File: api/src/controllers/account.controller.js [hunk 2]
```diff
@@ -42,33 +42,33 @@ const AccountController = {
...
-      });
-    }
-    // eslint-disable-next-line array-callback-return
-    return res.json({
-      status: 200,
-      data: allAccounts,
-    });
-  },
+  // fetchAllAccounts(req, res) {
+  //   const allAccounts = AccountService.fetchAllAccounts(req);
```
Comment (ucheg6) [inline api/src/controllers/account.controller.js]: Please no commented out codes

File: api/src/controllers/account.controller.js [hunk 3]
```diff
@@ -2,8 +2,8 @@ import AccountService from '../services/account.service';
 
 
 const AccountController = {
-  createAnAccount(req, res) {
-    const createdAccount = AccountService.createAccount(req.userId, req.body.type);
+  async createAnAccount(req, res) {
+    const createdAccount = await AccountService.createAccount(req.userId, req.body.type);
     if (createdAccount.error) {
       return res.json({
         status: 401,
```
Comment (ucheg6) [inline api/src/controllers/account.controller.js:9]: you have errors without error messages. Please add descriptive error messages

File: api/src/controllers/user.controller.js [hunk 4]
```diff
@@ -1,51 +1,81 @@
...
+  // },
 
-  addAUser(req, res) {
+  async addAUser(req, res) {
     const newUser = req.body;
     if (!newUser.type && !newUser.isAdmin) {
       newUser.type = 'client';
       newUser.isAdmin = false;
     }
+    if (!newUser.firstName) {
```
Comment (ucheg6) [inline api/src/controllers/user.controller.js]: can this be done with express validator or joi and it should be on the route. Write a validation middleware and call it on the route and not on the controller.

File: api/src/controllers/user.controller.js [hunk 5]
```diff
@@ -1,51 +1,81 @@
...
+    }
     const hashedPassword = bcrypt.hashSync(newUser.password, 8);
     newUser.password = hashedPassword;
-    const createdUser = UserService.addUser(newUser);
-    const token = jwt.sign({ id: createdUser.id }, secret, {
-      expiresIn: 86400,
-    });
-    createdUser.token = token;
+    const createdUser = await UserService.addUser(newUser);
     Object.defineProperty(createdUser, 'password', {
```
Comment (ucheg6) [inline api/src/controllers/user.controller.js]: This is unnecessary as you can decide the data to be returned from the query

File: api/src/controllers/user.controller.js [hunk 6]
```diff
@@ -1,51 +1,81 @@
...
   },
 
-  signIn(req, res) {
+  async signIn(req, res) {
     const oldUser = req.body;
-    const foundUser = UserService.signIn(oldUser);
+    if (!oldUser.email) {
+      return res.json({
+        status: 400,
+        error: 'Email field is required',
```
Comment (ucheg6) [inline api/src/controllers/user.controller.js]: same here. write a validator middleware and call it on the route

File: api/src/controllers/user.controller.js [hunk 7]
```diff
@@ -63,24 +93,20 @@ const UserController = {
...
-  getSingleUser(req, res) {
-    const { id } = req.params;
-    const foundUser = UserService.getAUser(id);
-    return res.json({
-      status: 201,
-      data: foundUser,
-    }).status(200);
-  },
+  // getSingleUser(req, res) {
+  //   const { id } = req.params;
```
Comment (ucheg6) [inline api/src/controllers/user.controller.js]: No comments!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: api/src/database/database.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+if (process.env.NODE_ENV === 'DEV') {
+  DATABASE_URL = process.env.DEVPG;
+} else if (process.env.NODE_ENV === 'PROD') {
+  DATABASE_URL = process.env.MAINPG;
+} else {
+  DATABASE_URL = process.env.TESTPG;
+}
+
+if (!process.env.NODE_ENV) {
+  DATABASE_URL = 'postgres://ysfjltsj:nOu1tZuIJf0D5K6oehu26V44fYTZmEqu@isilo.db.elephantsql.com:5432/ysfjltsj';
```
Comment (ucheg6) [inline api/src/database/database.js]: can't the DATABASE_URL's value come from .env?

File: api/src/services/account.service.js [hunk 9]
```diff
@@ -1,94 +1,132 @@
...
-    return validAccounts;
-  },
+  // fetchAllAccounts({ userId }) {
+  //   const user = UserService.getAUser(userId);
+  //   if (user.type === 'client') return { error: 'Unauthorized user' };
+  //   const validAccounts = dummyData.accounts.map((singleAccount) => {
+  //     const newAccount = new Account();
+  //     newAccount.id = singleAccount.id;
+  //     newAccount.accountNumber = singleAccount.accountNumber;
+  //     newAccount.createdOn = singleAccount.createdOn;
```
Comment (ucheg6) [inline api/src/services/account.service.js]: 🙅‍♂️ 🙅‍♂️ .... REMOVE!!!!

File: api/src/services/account.service.js [hunk 10]
```diff
@@ -1,94 +1,132 @@
...
+      client.release();
+    }
   },
 
-  getAnAccount(accountNumber) {
-    const account = dummyData.accounts
-      .find(singleAccount => singleAccount.accountNumber == accountNumber);
-    return account || {};
-  },
+  // getAnAccount(accountNumber) {
```
Comment (ucheg6) [inline api/src/services/account.service.js]: comments!!!!

File: api/src/controllers/account.controller.js (hunk 3 above)
Comment (Nerocodes) [reply api/src/controllers/account.controller.js:9]: There are error messages. The error messages are coming from the services. That's how it is for the rest to

File: api/package.json (hunk 1 above)
Comment (Nerocodes) [reply api/package.json]: I've merged it into one file now.

File: api/src/controllers/account.controller.js (hunk 2 above)
Comment (Nerocodes) [reply api/src/controllers/account.controller.js]: Removed

File: api/src/controllers/account.controller.js (hunk 3 above)
Comment (Nerocodes) [reply api/src/controllers/account.controller.js:9]: I send the error messages as an object to the controller

File: api/src/controllers/user.controller.js (hunk 4 above)
Comment (Nerocodes) [reply api/src/controllers/user.controller.js]: Done. I used joi

File: api/src/controllers/user.controller.js (hunk 5 above)
Comment (Nerocodes) [reply api/src/controllers/user.controller.js]: I've refactored this now.

File: api/src/controllers/user.controller.js (hunk 6 above)
Comment (Nerocodes) [reply api/src/controllers/user.controller.js]: Done. I used joi

File: api/src/controllers/user.controller.js (hunk 7 above)
Comment (Nerocodes) [reply api/src/controllers/user.controller.js]: Removed

File: api/src/database/database.js (hunk 8 above)
Comment (Nerocodes) [reply api/src/database/database.js]: I've added the environment variables to travis now.

File: api/src/services/account.service.js (hunk 9 above)
Comment (Nerocodes) [reply api/src/services/account.service.js]: Removed

File: api/src/services/account.service.js (hunk 10 above)
Comment (Nerocodes) [reply api/src/services/account.service.js]: Removed

**Pull request merged**